OAuth: Fix role mapping from id token

seanson opened the pull request:
As part of the new improvements to JMESPath in #17149 a commit ensuring
Role and Email data could be queried from id_token was lost.

**What this PR does / why we need it**:
This fixes an issue where an OAuth2 flow returns a valid token_id causes Grafana to search the wrong byte array when querying for Role and Email paths.

**Which issue(s) this PR fixes**:
Fixes #20243

**Special notes for your reviewer**:
Tested successfully with AzureAD with an Application role with the following ini config:
```
[auth.generic_oauth]
allow_sign_up = true
allowed_organizations =
api_url =
auth_url = https://login.microsoftonline.com/${TENANT_ID}/oauth2/authorize
client_id = ${CLIENT_ID}
client_secret = ${CLIENT_SECRET}
email_attribute_path = roles[0]
enabled = true
name = Azure AD
role_attribute_path = roles[0]
scopes = openid email profile roles
team_ids =
token_url = https://login.microsoftonline.com/${TENANT_ID}/oauth2/token
```

seanson committed:
OAuth: Fix JMESPath query for token_id

As part of the new improvements to JMESPath in #17149 a commit ensuring
Role and Email data could be queried from id_token was lost.

Co-authored-by: Timo Wendt <timo@tjwendt.de>

torkelo commented:
Looks good, a unit test for UserInfo would be good to safeguard against this regression happening again.

seanson committed:
Adds a regression test case for id_token role parsing.

seanson commented:
Thanks @torkelo! I've added a test case to cover the case of a valid id_token and role path.

seanson committed:
Fix linting issue

marefr commented:
@seanson thanks. Would be nice adding a test for the other case, i.e. when the token doesn't include an `id_token`. Maybe need to use `httptest` package to mock http request for user info then? Let us know if you need help and we can assist with this test.

seanson committed:
Add additional test for pulling role from apiUrl

seanson commented:
Thanks for the feedback @marefr! I've added a few more test cases, including one for mocking the user info. Let me know if this covers the regression enough.

yinzara commented:
I think you need to add the following into the `extractToken` method.  This ensure the api_url is still called if there is a defined "role_attribute_path" but it did not resolve to a valid role in the id_token:
```
    if role := s.extractRole(data, payload); s.roleAttributePath != "" && role == "" {
        s.log.Debug("No role found in id_token when role_attribute_path was configured", "json", string(payload), "data", data)
        return []byte{}, false
    }
```

marefr commented:
@yinzara that should already be handled by https://github.com/grafana/grafana/pull/20300/files#diff-4225a06af67a76978016c45030173d49R236

Or am I missing something here?

seanson committed:
* Add condition and test for handling valid id_token without a role
* Consolidate and simplify tests

seanson commented:
I've reproduced the issue @yinzara is talking about where a valid `id_token` without a matching role will prevent calling the apiUrl and added the fix and a test case to cover it: https://github.com/grafana/grafana/blob/669aa91f104e8848725aa75afa818dd64996daca/pkg/login/social/generic_oauth_test.go#L211

I've also refactored the test case to include both id_token and apiUrl testing as it was a lot easier to isolate the test case with both configs available.

yinzara reviewed (approved):
This all looks good to me.  proper test cases as well

marefr commented on pkg/login/social/generic_oauth.go:
```diff
@@ -294,6 +294,11 @@ func (s *SocialGenericOAuth) extractToken(data *UserInfoJson, token *oauth2.Toke
 		return []byte{}, false
 	}
 
+	if role := s.extractRole(data, payload); s.roleAttributePath != "" && role == "" {
```
This change doesn't make sense to me. What if you want to extract email address from token and above email check succeeds, but then this extraction of role doesn't find any role. Result will be that you will maybe get a role from UserInfo, but no email.

For now it's either extract from token or UserInfo, not both. Quite big refactoring is needed to support this so would suggest open a feature request for this specific use case and handle that change in a separate PR.

seanson replied:
@marefr I'll see if I can put together a test case for this so we can ensure we don't have a regression on email flow. I think without this specific line change we could definitely get an improvement but I agree, there's a good chunk of refactoring to be done.

marefr replied:
Okay, let me know.when you have something ready or want feedback. 

One idea for refactoring/simplify this may be to store token payload bytes on the UserInfoJson struct. 

Above change probably needs something along the lines:
 - If email or role are not found in token, user info should be inspected
 - if both email and role are found in token user info should not be inspected

henninge replied:
Hi @seanson  and @marefr !

I have looked into this and prepared a branch based on this one that picks up @marefr 's suggestions and should complete the fix. It is not a trivial fix so I can see how @seanson  might have gotten stuck on completing this. I tried to stick to the objective, though, and keep the fix as small as possbile.

I am not sure how to go from here. @seanson I think it would be easiest if you simply added my commits to this PR if you agree with the fix. Otherwise I could also open a new one. Please let me know what you think, I'd really like to see this issue fixed.

Thanks. ;-)

https://github.com/henninge/grafana/commits/17149-fix-oauth-data

seanson replied:
@henninge - Thanks for the help! Always glad for a helping hand. I'll merge this commit into this PR and write a test case to cover @marefr's flow.

seanson replied:
Changes rebased against master with the main auth cherry-picked and my refactored tests added. Should be good for review now. @henninge please ensure you've signed the CLA! :)

henninge replied:
Cool! Did you notice that there were two commits? I had refactored the tests, too. I think I came up with a couple extra test cases.

seanson replied:
@henninge - I had missed the extra test cases as we had both added the email/role checks. I've merged the new tests in with the existing test cleanup in the new commit.

I believe this should be good for review now.

marefr reviewed (changes requested):
See comment

cmanzi commented:
Thanks for working on this @seanson, i was running into the same issue. I created another PR(https://github.com/grafana/grafana/pull/20396) to address this before I saw yours - however, this solution seems to contradict the current [documentation](https://grafana.com/docs/auth/generic-oauth/#generic-oauth-authentication) if I'm reading it correctly. The docs state that the `role_attribute_path` will be evaluated against the response from the UserInfo endpoint, not the idToken. 

What if instead we only attempt to retrieve a role from the UserInfo output if a `role_attribute_path` is specified, to remain consistent with current docs?

seanson commented:
@cmanzi I would rather update the documentation to reflect the intent of the original code, which did expect to be able to pull roles from id_tokens.

yinzara commented:
> @cmanzi I would rather update the documentation to reflect the intent of the original code, which did expect to be able to pull roles from id_tokens.
I agree with you @seanson

sxd commented:
Why are the roles pulled from the id_token? and not in both id_token and access_token? On the other hand, the OpenID Connect specs doesn't say that the id_token should have other fields than the expected and never talks about having "roles" inside[1].

I think that the best idea it's to get the roles from the access_token since the standard implementation will put the roles/groups in the access_token and not in the id_token

1: https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest

seanson commented:
5.1.2 states that there are "standard" claims that are common names and "private" names that are arbitrary, which is why we allow pulling from id_token and an arbitrary path. Not allowing this would break common flows that provide this information in the id_token like Azure. 

I'm going to try and find some time this week to come around back to this PR as it's unfortunately a very busy time for me. 

>>> While this specification defines only a small set of Claims as standard Claims, other Claims MAY be used in conjunction with the standard Claims. When using such Claims, it is RECOMMENDED that collision-resistant names be used for the Claim Names, as described in the JSON Web Token (JWT)Jones, M., Bradley, J., and N. Sakimura, “JSON Web Token (JWT),” July 2014. [JWT] specification. Alternatively, Private Claim Names can be safely used when naming conflicts are unlikely to arise, as described in the JWT specification. Or, if specific additional Claims will have broad and general applicability, they can be registered with Registered Claim Names, per the JWT specification.

sxd commented:
@seanson Crap so Azure put the information inside id_token :/ well I think that getting the data from both should be ok. Thanks for your work!! =)

yinzara commented:
I think @sxd has a point that we should be able to configure grafana to look in the access token to find the role.  Maybe there should be a new configuration parameter of where it runs the role_attribute_path JMESPath query and allow for fallback.  

I propose, "role_attribute_source"

It should allow either "access_token", "id_token", or "api_url" as values and should support a list syntax (space separated) and attempt to resolve the role in the order specified.

Default value should probably be "access_token id_token api_url" (though I'd entertain other options).

That would allow us very configurable resolution.

yinzara commented:
Another feature I'd really like to see related to the "role_attribute_path", there should be a way of refusing a login (it shouldn't just default to the Viewer role) if the role returned from the role_attribute_path matched a specific value (maybe 'Deny')? 

I would like my oauth system to be able to properly authenticate a user but if that user didn't have the proper role, it shows the user an error message indicating they need to contact a system administrator to request access to the system.

Right now I've written a custom rule in Auth0 that if the user requests the grafana scopes but doesn't have access, it fails the login with an error message.

marefr commented:
@yinzara thanks for the input.

> I think @sxd has a point that we should be able to configure grafana to look in the access token to find the role (that's where Auth0 would put the roles). Maybe there should be a new configuration parameter of where it runs the role_attribute_path JMESPath query and allow for fallback.
> 
> I propose, "role_attribute_source"
> 
> It should allow either "access_token", "id_token", or "api_url" as values and should support a list syntax (space separated) and attempt to resolve the role in the order specified.
> 
> Default value should probably be "access_token id_token api_url" (though I'd entertain other options).
> 
> That would allow us very configurable resolution.

I definitely say no to more configuration settings. The whole reason using JMESPath is to limit the amount of configuration settings needed. 

Where did reading stuff from access token come from? That's a new feature and would suggest open an issue for that describing the use case behind it.

> Another feature I'd really like to see related to the "role_attribute_path", there should be a way of refusing a login (it shouldn't just default to the Viewer role) if the role returned from the role_attribute_path matched a specific value (maybe 'Deny')?

#17390

As a general rule of thumb we want to have small PR's focusing on solving one thing. That will dramatically increase the likelihood of it getting accepted and merged.

Still haven't received any response on https://github.com/grafana/grafana/pull/20300#discussion_r348608325

yinzara commented:
So for Auth0, I have to write a custom rule for Grafana because the only place the "roles" or "scopes" are present is in the access token. The id_token (and the response from the /userinfo endpoint) does not include the scopes.  The "role_attribute_source" would allow for that.  Technically we don't really NEED the configuration.  If we just changed the code to try the access token first, then the id token, then the api_url response in that order, we'd cover pretty much every use case.

I'll see if I can put together a PR for that one small piece of functionality.

sxd commented:
@yinzara actually I think the same you're right and that will cover all the aspect and using jmespath I think that will cover any kind of configuration =) so yes! I think that's the best approach! thank you for your work! thanks again for doing it that fast! kudos for you!

yinzara commented:
I think there may be confusion as to what I'm going to implement lol.  I don't have access to this branch so I can't make changes to it.  I was saying I'll develop the changes to make "Deny" returned from role_attribute_path evaluation refuse the login.

@seanson will still need to make the changes here.

JackIngleton commented:
Hi, are there any updates on this? I think that I'm experiencing the same issue when trying to use AWS cognito groups in the `role_attribute_path` setting.

seanson commented:
I'm back in the office this week and will hopefully have some time to pick it up.

henninge committed:
Refactor to extract from token and api consistently.

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=20300) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=20300) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: seanson<br/>:x: henninge<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=20300) it.</sub>

seanson committed:
Merge new generic OAuth tests with cleanup

Co-Authored-By: Henning Eggers <henning.eggers@inovex.de>

marefr commented:
Thanks. WIll have a look

marefr committed:
Merge remote-tracking branch 'origin/master' into 17149-fix-oauth-data

marefr committed:
Chore: Cleanup

marefr committed:
Remove GoConvey and use the standard library

marefr committed:
Restructure generic oauth

marefr reviewed (approved):
LGTM Great work everyone! To me it works as expected using Auth0 and Bitbucket. Haven't tested with Azure AD though, but given the great test coverage I feel confident to merge this.

Pushed some additional cleanup and restructure.

marefr commented:
@seanson @twendt @henninge Thank you for contributing to Grafana!

As soon as this have been built in master you should be able to use a nightly build to access this feature. Grafana v6.6.0-beta1 will hopefully be released on Monday and include this.

Thank you to everyone else that commented/provided feedback.

DanCech commented:
Nice work!

yinzara commented:
This is now available in the 6.0.0 release (which was released 3 days ago). Just thought everyone should know!
